### DISCUSSION TITLE
Make href attribute of next/link optional

### DISCUSSION THREAD

**Discussion by 314systems** [vercel/next.js#47698] [Ideas]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.3.0: Mon Jan 30 20:38:43 PST 2023; root:xnu-8792.81.3~2/RELEASE_ARM64_T8112
    Binaries:
      Node: 18.15.0
      npm: 9.5.0
      Yarn: N/A
      pnpm: 7.27.0
    Relevant packages:
      next: 13.2.5-canary.20
      eslint-config-next: 13.2.5-canary.20
      react: 18.2.0
      react-dom: 18.2.0
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

Routing (next/router, next/navigation, next/link)

### Link to the code that reproduces this issue

https://stackblitz.com/edit/vercel-next-js-zh6ppe?file=app/page.tsx

### To Reproduce

Write code that has no href attribute.

```jsx
{/* Error: Failed prop type: The prop `href` expects a `string` or `object` in `<Link>`, but got `undefined` instead. */}
<Link>without href</Link>
```

### Describe the Bug

The href attribute on `a` element is not required but `Link` requires the href attribute (See [HTML Standard § 4.6.2 Links created by a and area elements](https://html.spec.whatwg.org/multipage/links.html#attr-hyperlink-href)).

```jsx
{/* It's valid. */}
<a href="/">with href</a>

{/* It's also valid! */}
<a>without href</a>

{/* But, it's not valid... */}
<Link>without href</Link>
```

### Expected Behavior

The href attribute is not required.

```jsx
{/* href is optional */}
<Link>without href</Link>
```

**Comment by icyJoseph** [comment]
What would be the use case for `next/link`, without `href`?, you can still use a regular `a`nchor tag if you wish to, right?

To put it in perspective, should one be able to use the `react-router-dom` Link element without the `to` prop? For the correct usage of the framework, the `to` prop is required.

Another aspect would be that the `next/link` element can also take an object as href:

```jsx
      <Link href={{href: '/', query: {foo: 'bar'}}}></Link>
```

So it is not strictly following the HTML standard either. 

Link elements from router frameworks, are needed to tell the Router what to render, while still rendering an anchor tag to the DOM, all because we want the SPA experience, but no event exists for navigation, so the framework exposes a Link element.

> The https://developer.mozilla.org/en-US/docs/Web/API/Navigation_API aims to solve the lack of an event for navigation

What usage does an empty anchor tag, have for the Router?

If you want to say that you'd add onClick handlers, then this a11y lint rule would disagree:

- https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/HEAD/docs/rules/anchor-is-valid.md

**Reply by diegohaz** [reply]
Not passing `href` is [how you disable a link](https://twitter.com/diegohaz/status/1685159682187399168). The `next/link` requires the `href` prop even with a runtime check, and people are wrongly falling back to using a button instead so they can dynamically disable it, which is not accessible.

Some examples:

- https://github.com/vercel/next.js/discussions/47781#discussioncomment-6581052
- https://twitter.com/qzcurious/status/1685208954639618049
- https://www.youtube.com/watch?v=SXmni_7B0r4

Changing from `Link` to `a` may not be an obvious solution for most people. Also, I don't see how this could be abstracted away without specific knowledge of `next/link`. The abstraction would return props without `href`, but the `next/link` component would still complain about it. The only way to abstract it is to render `Link` or `a` as part of the implementation, which is not an option for lower-level component libraries like [Ariakit](https://ariakit.org).

Next.js could make that easier by getting rid of this requirement or, at least, the runtime check.
